exercises/practice/state-of-tic-tac-toe/.meta/gen.go

```diff
@@ -0,0 +1,73 @@
+package main
+
+import (
+	"log"
+	"text/template"
+
+	"../../../../gen"
+)
+
+func main() {
+	t, err := template.New("").Parse(tmpl)
+	if err != nil {
+		log.Fatal(err)
+	}
+	var j js
+	if err := gen.Gen("state-of-tic-tac-toe", &j, t); err != nil {
+		log.Fatal(err)
+	}
+}
+
+// The JSON structure we expect to be able to unmarshal into
+type js struct {
+	Cases []struct {
+		Description string
+		Cases       []oneCase
+	}
+}
+
+type oneCase struct {
+	Description string
+	Property    string
+	Input       struct {
+		Board []string
+	}
+	Expected interface{}
+}
+
+func (o oneCase) Result() string {
+	s, ok := o.Expected.(string)
+	if !ok {
+		return ""
+	}
+	return s
+}
+
+func (o oneCase) Err() string {
+	m, ok := o.Expected.(map[string]interface{})
+	if !ok {
+		return ""
+	}
+	return m["error"].(string)
+}
+
+// template applied to above data structure generates the Go test cases
+var tmpl = `package stateoftictactoe
+
+{{.Header}}
+
+var testCases = []struct{
+	description string
+	board []string
+	expected string
+	expectedErr string
```

Comment:
I think `expectedErr` should be a boolean indicating if the function is expected to return an error or not. 

I'm not a huge fan of checking for exact error messages, because it requires the student to read the tests carefully every time. If we want to go this route, at least we should have error variables the student can then use in the solution.

But I think just checking for the presence/absence of an error is enough, and we don't need to dig deep into the exact message given. Good error messages can be something to be discussed in a mentoring session.

Reply:
No, there's no problem in not using all the information in canonical-data.